src/scales/scale.category.js

```diff
@@ -1,135 +1,135 @@
+	getPixelForValue: function(value, index) {
+		var me = this;
+		var offset = me.options.offset;
+		// 1 is added because we need the length but we have the indexes
+		var offsetAmt = Math.max((me.maxIndex + 1 - me.minIndex - (offset ? 0 : 1)), 1);
+
+		// If value is a data object, then index is the index in the data array,
+		// not the index of the scale. We need to change that.
+		var valueCategory;
+		if (value !== undefined && value !== null) {
+			valueCategory = me.isHorizontal() ? value.x : value.y;
+		}
+		if (valueCategory !== undefined || (value !== undefined && isNaN(index))) {
+			var labels = me.getLabels();
+			value = valueCategory || value;
+			var idx = labels.indexOf(value);
+			index = idx !== -1 ? idx : index;
+		}
 
-				return me.left + widthOffset;
-			}
-			var valueHeight = me.height / offsetAmt;
-			var heightOffset = (valueHeight * (index - me.minIndex));
+		if (me.isHorizontal()) {
+			var valueWidth = me.width / offsetAmt;
+			var widthOffset = (valueWidth * (index - me.minIndex));
 
 			if (offset) {
-				heightOffset += (valueHeight / 2);
+				widthOffset += (valueWidth / 2);
 			}
 
-			return me.top + heightOffset;
-		},
-		getPixelForTick: function(index) {
-			return this.getPixelForValue(this.ticks[index], index + this.minIndex, null);
-		},
-		getValueForPixel: function(pixel) {
-			var me = this;
-			var offset = me.options.offset;
-			var value;
-			var offsetAmt = Math.max((me._ticks.length - (offset ? 0 : 1)), 1);
-			var horz = me.isHorizontal();
-			var valueDimension = (horz ? me.width : me.height) / offsetAmt;
-
-			pixel -= horz ? me.left : me.top;
+			return me.left + widthOffset;
+		}
+		var valueHeight = me.height / offsetAmt;
+		var heightOffset = (valueHeight * (index - me.minIndex));
 
-			if (offset) {
-				pixel -= (valueDimension / 2);
-			}
+		if (offset) {
+			heightOffset += (valueHeight / 2);
+		}
 
-			if (pixel <= 0) {
-				value = 0;
-			} else {
-				value = Math.round(pixel / valueDimension);
-			}
+		return me.top + heightOffset;
+	},
+
+	getPixelForTick: function(index) {
+		return this.getPixelForValue(this.ticks[index], index + this.minIndex, null);
+	},
 
-			return value + me.minIndex;
-		},
-		getBasePixel: function() {
-			return this.bottom;
+	getValueForPixel: function(pixel) {
+		var me = this;
+		var offset = me.options.offset;
+		var value;
+		var offsetAmt = Math.max((me._ticks.length - (offset ? 0 : 1)), 1);
+		var horz = me.isHorizontal();
+		var valueDimension = (horz ? me.width : me.height) / offsetAmt;
+
+		pixel -= horz ? me.left : me.top;
+
+		if (offset) {
+			pixel -= (valueDimension / 2);
+		}
+
+		if (pixel <= 0) {
+			value = 0;
+		} else {
+			value = Math.round(pixel / valueDimension);
 		}
-	});
 
-	scaleService.registerScaleType('category', DatasetScale, defaultConfig);
+		return value + me.minIndex;
+	},
+
+	getBasePixel: function() {
+		return this.bottom;
+	}
+});
+
+module.exports = function(scales) {
```

Comment:
Can we call this `scaleService` to be clearer? (same in the other charts)

Can we `require` the `scaleService` in this file instead of making it a `function` parameter?

Reply:
> Can we call this scaleService to be clearer?

I prefer not because at some point we will rename `Chart.scaleService` to `Chart.scales` to be consistent with `Chart.plugins` and `Chart.animations`.

> Can we require the scaleService in this file.

Better as argument else it would create an explicit dependency not needed.

Reply:
Is it correct that the individual scales will not be importable if a function parameter is required? Will it be a goal to make the scales importable and if so what would the plan for that be?

Reply:
Maybe we should move the `scales.registerScaleType` calls to happen in `src/chart.js`. That would remove the unnecessary dependency. It would also stop the scales from being registered multiple times. Right now I think the scale type gets registered everytime we import the scale somewhere

We would need to export the default config for each scale in order to do that, but that seems like it'd be a fine change

Reply:
> Maybe we should move the `scales.registerScaleType` calls to happen in `src/chart.js`.

That's what I did in my first branch, however I think we may need to handle more logic to register a scale, for example registering an internal plugin to [handle global doughnut state](/chartjs/Chart.js/pull/5951#issuecomment-451428512) (which applies to the controller, but may be similar for scales). But I agree and actually also prefer this approach (we could still have a scale static `register` method to handle these cases).

Reply:
Done: `Scale._default` (static) is currently private because I think this will change so don't want to commit on that API.